### ISSUE TITLE
Click/Touch doesn't work on mobile via PhoneGap with 0.7.1, works with 0.4.2

### ISSUE THREAD

**Issue by ashes999** [craftyjs/Crafty#1043]
Hi,

I crafted a very, very simple minimal CraftyJS game -- something like (minus boiler-plate code):

```
Crafty.e('2D, Alpha, Color, Canvas, Mouse, Touch').attr({ "w": 48, "h": 48 }).bind('Click', function() { Crafty.audio.play(someId) });
```

I tried swapping `Click` with `Touch`. It works fine on the web, but when I shove it into a PhoneGap container and deploy it on my phone (OnePlus X), the click handler never fires.  I even added a ".color('blue")` to make sure that it's not just that audio's broken on my phone.

Is this something that's supposed to work, or is it something that's not supported?

I'm using CraftyJS 0.7.1 with PhoneGap 6.3.0 (I think)

**Comment by ashes999** [comment]
I also tried building it with CocoonJS (2.0.0 with WebView), and it reproduced the same behaviour.

**Comment by ashes999** [comment]
There's something strange going on here. When I use Chrome Developer Tools to remotely debug, I can click (on my desktop's screen display -- the debug window) and the click-handler fires. But no amount of tapping my phone screen files it.

Some of the demos from the Crafty home page work fine, and some don't.

Any ideas @starwed or @mucaho ? I'm really stumped.

Edit: This is interesting. I found [this demo](http://craftyjs.com/demos/isometric/) which uses Crafty 0.4.2. If I switch from 0.7.1 to 0.4.2 and bind to `click` instead of `Click`, it handles touch on my phone.

**Comment by starwed** [comment]
I've never used either of those frameworks, so I'm not sure what's going on.  We bind to click (and other mouse events) like this:

``` js
Crafty.addEvent(this, Crafty.stage.elem, "click", Crafty.mouseDispatch);
```

> If I switch from 0.7.1 to 0.4.2 and bind to click instead of Click, it handles touch on my phone.

The code for click, specifically, looks to be quite different in that version of Crafty.  Instead of binding to the click event, it triggers click [when a mouseup occurs](https://github.com/craftyjs/Crafty/blob/0.4.2/src/controls.js#L57-L63).

I think there are a couple things you could check here:
- Do "MouseUp" and similar events still work, or are they all broken?
- Does PhoneGap have any debugging utilities that let you see how events are propagating?

I also did some quick googling, it sounds like perhaps PhoneGap is trying to do something "clever" here to reduce click latency -- maybe something there is incompatible with how Crafty is binding the click event to the stage element?  I wasn't able to find anything in the phonegap docs that specifically mentioned the click event, but I didn't look for very long.  :)

**Comment by ashes999** [comment]
Looks like you may be on to something. I added bindings to `MouseDown` and `MouseUp`, and only the `MouseUp` one fires.

I'm as new to PhoneGap as you are (almost).

[This post on SO](http://stackoverflow.com/questions/13892263/how-do-you-use-mouse-events-in-phonegap-cordova) seems to suggesting adding handlers for `touchstart` and `touchend` events instead. But sending these to `Crafty.stage.elem` doesn't seem to do anything:

```javascript
Crafty.addEvent(this, Crafty.stage.elem, "touchstart", Crafty.mouseDispatch);
Crafty.addEvent(this, Crafty.stage.elem, "touchend", Crafty.mouseDispatch);
```

Sorry, any ideas?

**Comment by ashes999** [comment]
It looks like binding to `MouseDown` and `MouseUp` events work.

I can use this as a temporary work-around.

Thanks!

**Comment by starwed** [comment]
Glad there's a workaround!  

I think I read somewhere that the framework implements a tap/ontap event -- if so you could also try listening for that, then passing along a modified version of the event to the touch dispatch function.

**Comment by ashes999** [comment]
I'm not convinced that this is PhoneGap's problem, because the demos from the CraftyJS home page also exhibit the same behaviour.

Hence, I'm not convinced that a PhoneGap-specific solution will help.

For now, I'm okay with the work-around. I hope this is something that can be fixed in Crafty itself, but I don't know how.

I tried to implement something like "use Click first, and if it doesn't work, use MouseUp" but I couldn't figure out how to make it only execute once on platforms where everything works (eg. Chrome on desktop).

If you need help testing, I have a device that reproduces the problem consistently, and I'll be happy to test on it for you.

**Comment by heniotierra** [comment]
Touch component worked in earlier versions of phonegap (~2 years old version). They must have changed their implementation. I could try investigating this, if I had the time. :/

**Comment by mucaho** [comment]
So as starwed pointed out, old code looks like 
```js
//click must mousedown and out on tile
if(e.type === "mousedown") {
    this.down = closest;
    this.down.trigger('mousedown', e);
} else if(e.type === "mouseup") {
    //check that down exists and this is down
    if(this.down && closest === this.down) {
        this.down.trigger("click", e);
    }
    //reset down
    this.down = null;
}
```

whereas new code looks like
```js
lastEvent = Crafty.lastEvent;
...
// trigger click when it should be triggered
if (lastEvent !== null && lastEvent.type === 'mousedown' && newType === 'mouseup') {
    newType = 'click';
    simulatedEvent = document.createEvent("MouseEvent");
    simulatedEvent.initMouseEvent(type, true, true, window, 1,
        first.screenX,
        first.screenY,
        first.clientX,
        first.clientY,
        false, false, false, false, 0, e.relatedTarget
    );
    first.target.dispatchEvent(simulatedEvent);
}
```

If new code doesn't work there may exist two reasons for that as far as I know:
* either PhoneGap removes all (synthetically) triggered Mouse events, but an [SO answer that synthetically triggers mouse events](http://stackoverflow.com/a/6175148) suggests otherwise
* or some other events happen between "mousedown"(= the simulated "touchstart") and "mouseup" (= the simulated "touchcancel"/"touchend"/"touchleave"), maybe "mousemove" happens (the simulated "touchmove")

I've tried to work around this by looking at the position difference from last "mousedown" to current "mouseup" event and trigger a synthetic mouse event "click" if the difference is small. See #1121.  @ashes999 if you are still interested, you could try to build that PR and test it on your device.

The proper solution would be to handle it similarly to the old implementation. Trigger "Click" if "MouseDown" and "MouseUp" happened on the same entity. We wouldn't want to do that in general case, because then we would trigger "Click" event twice. (that's similar to the problem ashes has with the fallback implementation).  
A possible solution to this is to inform mouseDispatch that the current event is a simulated event, e.g. by setting the `button` property in `initMouseEvent` to a negative value. This way mouseDispatch would know that it needs to trigger a "Click" event.

**Comment by ashes999** [comment]
I'm not personally still interested, but if you want me to test out your changes on the same device (since I opened an issue for it), I can make time to do that.

Do I need to pull/build the `mucaho:refactor-input2` branch?

**Comment by mucaho** [comment]
> I'm not personally still interested, but if you want me to test out your changes on the same device (since I opened an issue for it), I can make time to do that.

That would be awesome, thanks! Although, I won't bother you with this intermediate solution. There is another issue I discovered, so once a solution for that is implemented, I would gladly come back to your offer to test the final implementation. Read on.

> The proper solution would be to handle it similarly to the old implementation. Trigger "Click" if "MouseDown" and "MouseUp" happened on the same entity.

There's actually another issue present with all `Click` and `DoubleClick` events on Canvas and WebGL. If you mouse down over the background, but then move your mouse and release it over a Canvas/WebGL entity with a click handler, it will incorrectly fire the `Click` on the entity. However, a `Click` should only occur if mousedown and mouseup happened on same entity.
```js
Crafty.e('2D, Canvas, Color, Mouse')
  .attr({x: 0, y: 0, w: 100, h: 100})
  .color('orange')
  .bind('Click', function() {
    console.log('Click');
  });
```
After #1121 is merged, I will try to streamline the pointer implementation and try to fix the bug mentioned in this comment and the bug above. The approach I have in mind is to listen to native Click and DoubleClick events, but fire only "valid" ones by keeping track of previous targets similar to how it was implemented in 0.4.2.

> A possible solution to this is to inform mouseDispatch that the current event is a simulated event, e.g. by setting the button property in initMouseEvent to a negative value. This way mouseDispatch would know that it needs to trigger a "Click" event.

(As a side note, mostly as a reminder to myself: It can even be simplified to always trigger Click after a pair of touchstart / touchend events occurred, no matter their delta position. MouseDispatch will take care of figuring out if Click happened on the same entity and will propagate both valid mouse clicks and touch clicks in that case).

**Comment by ashes999** [comment]
@mucaho if you want to "bother" me with intermediate solutions or trying multiple builds, I don't have any issues with that. I'd rather give you feedback on things you're not sure about, and get them working, than have you spend a bunch of time on a fix that ultimately doesn't resolve the problem.

It shouldn't be hard to revive my PhoneGap dummy/tester app, so let me know when and how to test -- I think I just need a `crafty.min.js`/`crafty.js` file (or build instructions on how to generate one).

**Comment by mucaho** [comment]
> if you want to "bother" me with intermediate solutions or trying multiple builds, I don't have any issues with that. I'd rather give you feedback on things you're not sure about, and get them working, than have you spend a bunch of time on a fix that ultimately doesn't resolve the problem.

Great, many thanks! Following is the link to the `crafty.js` file (masked as a .png). Just rename the file to crafty.js. If there are any problems, let me know!
![crafty.js](https://cloud.githubusercontent.com/assets/3935691/25784404/7eec3cac-336c-11e7-9e37-006af760f2a1.png)

**Comment by mucaho** [comment]
Created the hello-world example, added Crafty to it and ran in a web browser and through PhoneGap developer on my Android 6.0 phone.
Both old and new crafty versions seem to work fine with regards to click using these platforms. The newer version I linked above is a bit more forgiving when clicking though. However, it remains to be seen how these versions work when deployed on the phone in the "real" container. If you could test that, I would be grateful!

So we can compare results, I have put together an example test code. This test code should be put into the `<body>` tag of `www/index.html`, after [creating the Hello World example](http://docs.phonegap.com/getting-started/3-create-your-app/desktop/). Put the crafty library in `js/crafty.js`
EDIT: I'm not sure if Crafty will add the proper component when using the `Button` component (It tries to determine whether Crafty runs on mobile - I don't know if PhoneGap emulates that correctly). Try switching this component in the code below with `Mouse` and/or `Touch`, but don't use both as [in your initial example](https://github.com/craftyjs/Crafty/issues/1043#issue-167445649). It's either `Button`, `Mouse` or `Touch`, but not any combination of them.
EDIT2: `Button` works by checking `Crafty.mobile`, which in turn checks for `window.navigator.userAgent`. A [StackOverflow answer](http://stackoverflow.com/a/11078197) seems to suggest it should work checking for that property.
```html
<body>
    <div class="app">
        <h1>PhoneGap</h1>
        <div id="deviceready" class="blink">
            <p class="event listening">Connecting to Device</p>
            <p class="event received">Device is Ready</p>
        </div>
        <div id="game"></div>
    </div>
    <script type="text/javascript" src="cordova.js"></script>
    <script type="text/javascript" src="js/index.js"></script>
    <!-- add CraftyJS library -->
    <script type="text/javascript" src="js/crafty.js"></script>
    <script type="text/javascript">
        /** start game once device ready **/
        document.addEventListener('deviceready', function() {
            Crafty.init();
            Crafty.background('white');
            // toggle this flag
            Crafty.multitouch(false);

            // this box serves as a click handler for when multitouch is disabled
            if (!Crafty.multitouch()) {
                Crafty.e('2D, Canvas, Color, Button')
                    .attr({x:100, y:100, w:64, h:64, toggle: false, lastToggleTime: -Infinity})
                    .color('green')
                    .bind('Click', function() {
                        var timeNow = Date.now();
                        if (timeNow > this.lastToggleTime + 1000) { // ignore clicks within 1 second of each other
                            this.color(this.toggle ? 'green' : 'blue');

                            this.toggle = !this.toggle;
                            this.lastToggleTime = timeNow;
                        }
                    });
            // this box serves as a touch handler for when multitouch is enabled
            } else {
                Crafty.e('2D, Canvas, Color, Touch')
                    .attr({x:100, y:100, w:64, h:64, toggle: false, lastToggleTime: -Infinity})
                    .color('green')
                    .bind('TouchStart', function() {
                        this.color('blue');
                    })
                    .bind('TouchMove', function() {
                        this.w++;
                        this.h++;
                    })
                    .bind('TouchEnd', function() {
                        this.w = 64;
                        this.h = 64;
                        this.color('green');
                    });
            }

        }, false);

        /** initialize device **/
        app.initialize();
    </script>
</body>
```

**Comment by ashes999** [comment]
Sorry, it was always my intention to test this on my real device; I just didn't get time yet. It's my intention to get to this by the end of the week.

**Comment by ashes999** [comment]
After lots of hoopla related to network access, I can confirm that the touch event works as expected on my device.

I noticed that in both web (Chrome from desktop) and on my device, I have to (after the first click) click the square several times to get it to toggle -- but the first click always changes it from green to blue.

You can close this issue or leave it open as per your processes. As far as I'm concerned, problem solved!

**Comment by mucaho** [comment]
Many thanks for testing!
> I noticed that in both web (Chrome from desktop) and on my device, I have to (after the first click) click the square several times to get it to toggle -- but the first click always changes it from green to blue.

I added a timeout of 1s to the example, as Chrome receives two Click events. Otherwise the box would change from blue to green and back to green quickly, making it appear as if nothing happened.

Will leave this issue open, as a new implementation is coming soon, that will supersede this one. So we need to test again :/

**Comment by ashes999** [comment]
No worries. Just ping me / reply whenever you need me to re-test.
